### PR TITLE
[javascript] add npm-mode

### PR DESCRIPTION
**Pull request by hexagonal-sun**
Add [npm-mode](https://github.com/mojochao/npm-mode) to the javascript layer.  This allows easy execution of npm
commands at the project root from any javascript file in the project.

### CODE REVIEW

File: layers/+lang/javascript/packages.el [hunk 1]
```diff
@@ -65,6 +66,21 @@
...
+  (use-package npm-mode
+    :defer nil
+    :init
+    (progn
+      (add-hook 'js2-mode-hook #'npm-mode))
+    :config
+    (progn
+      (spacemacs/declare-prefix-for-mode 'js2-mode "mn" "npm")
+      (spacemacs/set-leader-keys-for-major-mode 'js2-mode
+        "ni" 'npm-mode-npm-install
```
Comment (lebensterben) [inline layers/+lang/javascript/packages.el:79]: How about adding other commands and keybindings?

https://github.com/mojochao/npm-mode/blob/3ee7c0bad5b7a041d4739ef3aaa06a3dc764e5eb/npm-mode.el#L34-L41

File: layers/+lang/javascript/README.org [hunk 2]
```diff
@@ -326,6 +327,16 @@ parse warnings:
 | ~SPC m z F~ | toggle hide functions    |
 | ~SPC m z C~ | toggle hide comments     |
 
+** Running NPM (npm-mode)
+
+| Key binding | Description                                                                     |
+|-------------+---------------------------------------------------------------------------------|
+| ~SPC m n i~ | Run the =npm install= command in the project root                               |
+| ~SPC m n r~ | Show a list of avilable npm scripts, and execute the selected one               |
```
Comment (smile13241324) [inline layers/+lang/javascript/README.org:335]: Small typo should be available

File: layers/+lang/javascript/packages.el (hunk 1 above)
Comment (smile13241324) [reply layers/+lang/javascript/packages.el:79]: Please also add the following missing commands for a full integration:
Command|Binding|Description
--------------|----------|---------------
npm-mode-npm-init             | <kbd>SPC m n n</kbd> | Initialize new project
npm-mode-npm-uninstall        | <kbd>SPC m n u</kbd> | Remove project dependency           
npm-mode-npm-list             | <kbd>SPC m n l</kbd> | List installed project dependencies 
npm-mode-visit-project-file   | <kbd>SPC m n p</kbd> | Visit project package.json file

**Review by smile13241324 — CHANGES_REQUESTED** [review]
I like the idea to get some of the NPM project maintenance work integrated into Spacemacs. However when we do that we should try to integrate all available commands to allow users to maintain their project without having to constantly switch between NPM and Spacemacs.

I have added a list of additional commands you should integrate prior to having this merged. Also I have found some smaller typos you should fix on the way.

File: layers/+lang/javascript/packages.el (hunk 1 above)
Comment (hexagonal-sun) [reply layers/+lang/javascript/packages.el:79]: Added commands as requested.  Thanks for the suggestion!

**Comment by hexagonal-sun** [comment]
@smile13241324  @lebensterben Thanks for the review!  Your comments should have been addressed. Please take another look and let me know.

**Comment by hexagonal-sun** [comment]
One quick question.  It'd be really useful to have these commands avilable in the TypeScript mode too.  Is there a way to have it in both layers without having to duplicate the code?

**Comment by lebensterben** [comment]
https://github.com/syl20bnr/spacemacs/blob/7276970c914af32f7e3d3f45e654eaa2e04bf81e/layers/%2Blang/typescript/layers.el#L12

typescript layer depends on javascript.

So if you define a function in javascript layer, and it should be callable in typescript layer.

**Comment by hexagonal-sun** [comment]
@lebensterben Ah, I didn't know there was an inheritence structure with layers.  Does that mean that the TypeScript mode will also inherit the javascript layer's keybindings?

**Review by smile13241324 — APPROVED** [review]
Looks good now

**Comment by smile13241324** [comment]
No this is not an inheritance structure, instead its a structure of dependencies i.e. when typescript-layer is loaded also the javascript-layer is loaded. You can derive modes from one another in emacs, though I have not made the best experiences with using this feature.

The best way to have a layer use a package introduced in another layer is by using the `pre-init-*` construct. In your case you would add this to the typescript layer which would only be executed if the package * is initialised in this function you would add the init code for typescript-mode or actually for the typescript modes, I think they have more than one.

More details can be found in the layers documentation, reachable from within spacemacs with <kbd>SPC h SPC layers return</kbd>

**Pull request merged**